### PR TITLE
Charstream implementation

### PR DESCRIPTION
**Pull request by Skallwar**

### CODE REVIEW

File: include/charstream.h [hunk 1]
```diff
@@ -8,8 +8,12 @@
  * Transform input from a file descriptor into a stream of characters.
  */
 
+#define CHARSTREAM_EOF EOF
```
Comment (CohenArthur) [inline include/charstream.h:11]:
```suggestion
#define CHARSTREAM_EOF (EOF)
```
EOF expands to a numerical value so it's always nice to have parentheses around those

File: include/charstream.h [hunk 2]
```diff
@@ -8,8 +8,12 @@
...
+#define CHARSTREAM_ERR (-2)
+
 struct charstream {
 	/* Empty for now, it will be filled as implementation is written. */
+	FILE *file;
```
Comment (CohenArthur) [inline include/charstream.h:16]: Remove comment regarding emptiness of struct

File: meson.build [hunk 3]
```diff
@@ -14,4 +14,15 @@ src = [
   'src/charstream.c',
   ]
 
+
 executable('ubsh', src, include_directories: inc)
+
+unit_tests_src = [
+  'src/charstream.c',
+  'tests/charstream.c'
+  ]
```
Comment (CohenArthur) [inline meson.build:23]: Maybe we could come up with something better to add every source file but main and the test source files? We can concatenate arrays with meson, that might be a way

File: src/charstream.c [hunk 4]
```diff
@@ -1,17 +1,36 @@
...
+
+	return chr;
 }
 
 char charstream_peek(struct charstream *cs)
 {
-	UNIMPLEMENTED();
+	char chr = charstream_read(cs);
+	if (chr != CHARSTREAM_ERR)
+		fseek(cs->file, -1, SEEK_CUR);
```
Comment (CohenArthur) [inline src/charstream.c:33]: What about `fgetc(); ungetc();` ? Easier to grasp imo

File: tests/charstream.c [hunk 5]
```diff
@@ -0,0 +1,51 @@
+#include <criterion/criterion.h>
+
+#include "charstream.h"
+
+#define STRING_EMPTY ""
```
Comment (CohenArthur) [inline tests/charstream.c:5]: No tests with `STRING_EMPTY`?

File: meson.build (hunk 3 above)
Comment (n1tram1) [reply meson.build:23]: Good idea, we should look into it

**Review by n1tram1 — APPROVED** [review]

File: include/charstream.h (hunk 1 above)
Comment (Skallwar) [reply include/charstream.h:11]: In the case of glibc EOF expands to ``(-1)`` [see](https://github.com/lattera/glibc/blob/895ef79e04a953cac1493863bcae29ad85657ee1/libio/stdio.h#L104)
But yes valid point

File: meson.build (hunk 3 above)
Comment (Skallwar) [reply meson.build]: Fixed

File: src/charstream.c (hunk 4 above)
Comment (Skallwar) [reply src/charstream.c]: Fixed 5e16e8c

**Review by n1tram1 — APPROVED** [review]

**Pull request merged**